### PR TITLE
vCard 4.0: fix Name Property

### LINKED ISSUE #113: Create method to check "This is the only mandatory property."
FN is mandatory, so if we want to parse, we need to check its existence

### PR DESCRIPTION
**Pull request by helveticadomes**
fix Name Property (wrong order: FN and N have different orders!) 
mandatory check for fullname (fix #113 ) based on #128

### CODE REVIEW

**Comment by helveticadomes** [comment]
@danger89 maybe you can help me 
- [ ] to pass travis checks
- [ ] review the mandatory check 
- [ ] review the mandatory exception
- [ ] confim that FN / N need that fix

**Comment by danger89** [comment]
Ah the name parser was also wrong. So the order was indeed wrong regarding lastname, firstname and additional. 

It's also a best practice to add some additional unit testers with introducing of new code.
Indeed the RFC 6350 says regarding FN: "The property **MUST** be  present in the vCard object."

While the N (Name) **SHOULD** be present.

Is is OK to assume that when N is filled-in, FN can be derived from that? So still ending up in the vCard, without explicitly setting this? Since, this will be my use case. With the option to set the FN manually whenever the user likes this to do. 

Eg. `N:Doe;J.;;;` generates FN for you if not specified to the vCard object: `FN:J. Doe`. However, when both FN & N is not filled-in, only then return an error.

**Comment by danger89** [comment]
@helveticadomes  I changed my unit tester on dev branch, can you pull the new changes from 2.0.0-dev? So I can validate the FN outcome.

**Comment by helveticadomes** [comment]
> Is it OK to assume that when N is filled-in, FN can be derived from that? 

for me that make most sense, or do you think i should not do that?

**it is implemented that:** 
if FN exist and N exist => keep both and do nothing to them
if FN exist but N dont exist => keep FN and do nothing with N
if FN & N dont exist => throw exception that FN is mandatory
if FN dont exist but N exist => silent create FN from N, and add it to the vcard

pull which changes where and how?

**Comment by danger89** [comment]
> **it is implemented that:**
> if FN exist and N exist => keep both and do nothing to them
> if FN exist but N dont exist => keep FN and do nothing with N
> if FN & N dont exist => throw exception that FN is mandatory
> if FN dont exist but N exist => silent create FN from N, and add it to the vcard

That sounds good indeed!


> pull which changes where and how?

You already did, how to trigger a rebuild on this PR :confused: (lack of knowledge of Github <> Travis integration I admit):
https://github.com/helveticadomes/vcard/commit/4220a451e4b91de24b49795e4103f67f371a75ad

**Comment by danger89** [comment]
is it a good idea to also **add** a correct vcard file to assets? With correct I mean define the `FN:`.
https://github.com/helveticadomes/vcard/blob/feature-fixed-name-property/tests/assets/vcard.vcf

We can keep the current vcard asset as well (rename it?), since I think its a good test to validate if the jeroendesloovere/vcard parser is robust enough for parsing vcards files without FN (and only N) -> although its is an invalid Vcard file I think the parser should be robust.

File: src/VCard.php [hunk 1]
```diff
@@ -160,9 +160,38 @@ public function getParameters(string $filterByPropertyParameterClass = null): ar
...
+                } else {
+                    $others[] = $value;
+                }
+            }
+            //check for empty and throw exception, if it can not be generated by existing fields
+            if (count($found_fullname) == 0) {
+                if (count($found_name) == 0) {
+                    throw VCardException::forRequiredProperty(new FullName('NoName'));
+                } else {
+                    $found_fullname[] = new FullName(implode(' ', array_filter(array($found_name[0]->getPrefix(), $found_name[0]->getLastName(), $found_name[0]->getAdditional(), $found_name[0]->getFirstName(), $found_name[0]->getSuffix()))));
```
Comment (danger89) [inline src/VCard.php]: Looking at the result of the fullname of the unit tester ran https://travis-ci.org/jeroendesloovere/vcard/jobs/502040710.

`+FN:Mr. van den Berg Antoine Melroy\r\n`

This is not correct.

I prefer:  Prefix - firstname - additional - lastname - suffix order.

Which should result into:
`+FN:Mr. Melroy Antoine van den Berg\r\n`

I hope you agree as well regarding this.

**Comment by helveticadomes** [comment]
lack of knowledge that describe exactly me :-D (git/github/travis)
if i assume right something need to be added to VCardTest.php because FN is added if only N exist.
but with all this dependencies it is something that begin slightly overstrain my capability

File: src/VCard.php (hunk 1 above)
Comment (helveticadomes) [reply src/VCard.php]: i agree, fied that & forcepush was to remove the commits createt form a wron rebase

**Comment by danger89** [comment]
> overstrain



> 
> 
> lack of knowledge that describe exactly me :-D (git/github/travis)
> if i assume right something need to be added to VCardTest.php because FN is added if only N exist.
> but with all this dependencies it is something that begin slightly overstrain my capability

No worries. I can either do it myself. Or better, guide you through the process to help you. So next time you do have the ability and skills to do it yourself :raised_hands: . I like the second approach, and help you.

File: src/VCard.php [hunk 2]
```diff
@@ -160,9 +160,38 @@ public function getParameters(string $filterByPropertyParameterClass = null): ar
...
+                } else {
+                    $others[] = $value;
+                }
+            }
+            //check for empty and throw exception, if it can not be generated by existing fields
+            if (count($found_fullname) == 0) {
+                if (count($found_name) == 0) {
+                    throw VCardException::forRequiredProperty(new FullName('NoName'));
+                } else {
+                    $found_fullname[] = new FullName(implode(' ', array_filter(array($found_name[0]->getPrefix(), $found_name[0]->getFirstName(), $found_name[0]->getAdditional(), $found_name[0]->getLastName(), $found_name[0]->getSuffix()))));
```
Comment (to-be-the-best) [inline src/VCard.php]: Can this be made more readable, for example something like:

```php
$found_fullname[] = new FullName(implode(' ', array_filter(array(
    $found_name[0]->getPrefix(), 
    $found_name[0]->getFirstName(), 
    $found_name[0]->getAdditional(), 
    $found_name[0]->getLastName(), 
    $found_name[0]->getSuffix(),
))));
```

**Comment by danger89** [comment]
I fixed the testcases, found yet another bug in the `DateTimeOrStringValue` class. regarding the `__toString()` method.

The 'u' format (see Date format in php docs) will return zero's with a date! So that is definitely not the behavior you want. 

Therefore I urge you to also look to the `DateTimeValue` which properly also need the same fix as `DateTimeOrStringValue`?

I also created a `composer test`, `composer standard` and finally a `composer fix-standard`. You're welcome :+1: .

File: composer.json [hunk 3]
```diff
@@ -23,5 +23,10 @@
     },
     "autoload": {
         "psr-4": { "JeroenDesloovere\\VCard\\": "src/" }
+    },
+    "scripts": {
+        "test": "phpunit tests",
+	"standard": "phpcs --standard=psr2 --extensions=php --warning-severity=0 --report=full 'src'",
```
Comment (to-be-the-best) [inline composer.json]: Please use spaces not tabs just like the rest of the file

Comment (danger89) [reply composer.json]: ow oopsy thx

Comment (to-be-the-best) [reply composer.json]: Your welcome

File: src/VCard.php (hunk 2 above)
Comment (danger89) [reply src/VCard.php]: I will fix this for helveticadomes.

File: composer.json (hunk 3 above)
Comment (danger89) [reply composer.json]: Lessons learned: never use Gedit, Atom & vim all together :rofl:

File: tests/VCardTest.php [hunk 4]
```diff
@@ -32,32 +33,48 @@
...
 {
     /** @var VCard */
-    private $firstVCard;
+    protected static $firstVCard;
 
     /** @var VCard */
-    private $secondVCard;
+    protected static $secondVCard;
+
+    /** @var VCards: containing both first & second vcard together */
```
Comment (to-be-the-best) [inline tests/VCardTest.php]: should this not be of the type VCard?
https://github.com/jeroendesloovere/vcard/pull/146/files#diff-fc15453e7e481d3e92f86a7d1d2e7a99R101 here it's set with VCard

Comment (to-be-the-best) [reply tests/VCardTest.php]: Also i'm not sure, but i don't believe the : is needed/allowed in the phpdoc specs

Comment (danger89) [reply tests/VCardTest.php]: ya ok its type Vcard.. I wanted to point out this variable stores two vcards together.

I don't know what I was thinking.. I will revert it.

**Comment by danger89** [comment]
@jeroendesloovere I think it's ready to be merged.

File: src/Property/Value/DateTimeValue.php [hunk 5]
```diff
@@ -16,6 +16,7 @@ public function __construct(\DateTime $value)
 
     public function __toString(): string
     {
+        // TODO: possibly not in line with rfc6350
```
Comment (danger89) [inline src/Property/Value/DateTimeValue.php]: However, somebody could also look to this toString() method. However, this can be done later as well (has nothing to do with the Name property fix)

**Comment by danger89** [comment]
@helveticadomes Do you like to join the Gitter chat? https://gitter.im/vcard-php-library/community
This way we can easier help each other.

**Comment by danger89** [comment]
Ps.
@helveticadomes  I don't get the "FullName is mandatory" exception when I only use the `getParameters()`. But only with `getContent()`.

File: src/Property/Value/DateTimeValue.php (hunk 5 above)
Comment (danger89) [reply src/Property/Value/DateTimeValue.php]: Fixed. Thank you danger89 :+1:

File: src/Property/Value/DateTimeOrStringValue.php [hunk 6]
```diff
@@ -17,7 +17,8 @@ public function __construct($value)
     public function __toString(): string
     {
         if ($this->value instanceof \DateTime) {
-            return $this->value->format('u');
+            // According to DATE-AND-OR-TIME rfc6350 standard
+            return $this->value->format('Ymd\THis');
```
Comment (to-be-the-best) [inline src/Property/Value/DateTimeOrStringValue.php]: Does this also need the `z` just like https://github.com/jeroendesloovere/vcard/pull/146/files#diff-0a46c638422a54cb3e31646555c0c263R20

Comment (to-be-the-best) [inline src/Property/Value/DateTimeOrStringValue.php]: I also want to raise a point for discussion, I see that in rfc6350 all the references to date and/or time basically say to use ISO 8601.
The thing is that php has a constant for ISO 8601 (DateTimeInterface::ATOM) which has the following syntax: Y-m-d\TH:i:sP (example: 2005-08-15T15:52:01+00:00), but this is different from the examples in rfc6350.
Personally i prefer the DateTimeInterface::ATOM style.
I'm interested in your opinions.

**Comment by helveticadomes** [comment]
> @helveticadomes I don't get the "FullName is mandatory" exception when I use the `getParameters()`. But only with `getContent()`.
if you use getParameters() instead of getProperties() it is clear that you would not get the FullName exception ;-)

PS: i assume we need to add this expection also to VCardTest.

**Comment by to-be-the-best** [comment]
> > @helveticadomes I don't get the "FullName is mandatory" exception when I use the `getParameters()`. But only with `getContent()`.
> 
> if you use `getParameters() `instead of `getProperties() `it is clear that you would not get the FullName exception ;-)
> 
> PS: i assume we need to add this expection also to VCardTest.

Indeed the expection needs tests

**Comment by danger89** [comment]
Currently its tested indirectly with several testcase like: https://github.com/jeroendesloovere/vcard/blob/9985584fdadd15761679e6e312719db0cf099d44/tests/VCardTest.php#L353

As you can see these tests are **not** resulting into an exception/error. But indeed you want to have a test that proves the exception is actually trigger also (bad weather test).

**Comment by danger89** [comment]
Ready to merge now..? Please say yes :heart:

File: src/Property/Value/DateTimeOrStringValue.php (hunk 6 above)
Comment (danger89) [reply src/Property/Value/DateTimeOrStringValue.php]: No, without Z:
https://tools.ietf.org/html/rfc6350#page-14

Comment (danger89) [reply src/Property/Value/DateTimeOrStringValue.php]: @to-be-the-best The RFC6350 says it's BASED ON that ISO standard: and that ISO standard mainly says this is the standard ISO notation: **YYYY-MM-DD**. Bear in mind that the ISO standard is very brought, and is not specific enough. That is why the RFC6350 is extended this ISO standard, to be more standard for Internet Applications.

I'm just looking at the specification of that document:

> 
> To allow unambiguous
>    interpretation, a stand-alone TIME value is always preceded by a "T".
> 
>    Examples for "date-and-or-time":
> 
>              19961022T140000

So I think the `Z` is debatable indeed. 

And the [ISO 8601](https://www.iso.org/standard/40874.html) is even revised.  And have two different versions `ISO/DIS 8601-1` & `ISO/DIS 8601-2`.

TLDR;To conclude: `Ymd\THis` is OK according to the RFC6350 spec. The RFC standard is based on the ISO standard, but extends that, therefor I like to look to RFC standard only. Since RFC6350 is leading.
Finally, I think '\Z' can be added if you wish.

Comment (danger89) [reply src/Property/Value/DateTimeOrStringValue.php]: Ps. plus we need to do manual test procedures to validate if this notation is indeed working for the most used applications (eg. Outlook, Google Contacts, iCloud, ..)

Comment (danger89) [reply src/Property/Value/DateTimeOrStringValue.php]: Ps. ps. Actually another related discussion I like to have instead, is how to define a PHP class that can be used to only output `Ymd`? Since the PHP `DateTime()` can be used to store both date & time. So I can only assume now that there WILL be a time included (also 00:00:00 could even be valid) and dump the full date time RFC standard notation into my vcard. 

However, it would be nice to have an option to only a date (without time) and time (without date), with the proper toString() implementation for those two cases (together with the full date time case-> already implemented).

**Comment by danger89** [comment]
TODO: Create issue regarding DateTime to support also a single date or time object (with a proper toString() implementation for these two cases). Currently the DateTime can store both a date and time (00:00:00 is a valid time), therefor it always output the full date + time to the vcard currently.

However Google Contacts can't handle this actually, despite the RFC standrd. Regarding the birthday:
`BDAY:19891207` results into (GOOD):
image


`BDAY:19891207T104310` Results into (Not good :sob: , despite the standard followed):
image
